### PR TITLE
code executing in webgui

### LINKED ISSUE #144: #Enhancement - Implement run code via webgui
I'm not into typescript and don't really know where to get started if I was to help with this.

But thought it would be cool if you could test your developments directly from VS code instead of opening the GUI. You could add a flag in the settings on how you want it to be executed if the webgui isn't enabled on the abap system.

But you can call SE38 in the webgui for example by
https://localhost:50001/sap/bc/gui/sap/its/webgui?~transaction=*SE38%20RS38M-PROGRAMM=ZTEST;DYNP_OKCODE=STRT#...
This executes the program automatically.

### PR DESCRIPTION
**Pull request by uxkjaer**
Fixes #144

### CODE REVIEW

**Comment by uxkjaer** [comment]
still a bit to go. I've added the config settings and some sample. Not really sure how to get the url, type of object etc to continue from here. 
Any help @marcellourbani ?

**Comment by marcellourbani** [comment]
@uxkjaer 
Thanks a lot for the contribution
There are a few things who don't work for me though:

1. boolean parameters don't work, there should be a 3 option dropdown going like:
    1. SAPGUI(default)
    2. webgui in  browser
    3. webgui in vscode
2. no need for a case statement. Transaction  **SADT_START_WB_URI** used by  **showInGui** does any object type
3. the code to handle the browser goes in function **runInSapGui**, in sapgui.ts around line 44
4. I will bump version  when I publish the next release

Should look more or less like this:
image

Thank you
Marcello

**Comment by uxkjaer** [comment]
The SADT_START_WB_URI transaction only loads the program, it doesn't execute it. My intention was to execute the program directly so we don't have to "run" it in the webgui. 

What do you think?

**Comment by marcellourbani** [comment]
Oh yes, you're right!
I think we need a new command to open&run to provide what you're after.
I also think the code above should stay more or less as it is, is the command  which should be different for running.
Sorry I got confused, the fact is we are talking about 2 different features:

1. use the webgui in place of sapgui
2. run instead of open

So please also add a new command for this. Changes are pretty trivial:

- new command constant, say executeImmediately
- @command entry in its main method
- matching command in package.json

Would also consider calling this execute and renaming the old command from execute to openingui or similar, but might break stuff

**Comment by uxkjaer** [comment]
i've implemented the code to execute and then changed the existing to openingui. I can't really figure out how to define the command in the command pallette. Do i need to add it to the package.json file. I just see other commands in there that aren't present either. 

I do have one issue for executing on class. the file.object.name contains both the classname and method. I can't figure out where that is added. When I debug the AbapObjectBase it looks correct. Do you have an idea?
image

**Comment by marcellourbani** [comment]
They're defined in package.json:
image
and in structure AbapFsCommands

**Comment by uxkjaer** [comment]
Yup, that's sorted now. Last issue is that of the class name having the method on it as well.

**Comment by marcellourbani** [comment]
You can get the object from the url. Plenty of examples around

**Comment by uxkjaer** [comment]
Yeah it's just that all other objects works with the name. But I'll change it to the url.

**Comment by uxkjaer** [comment]
I've finished the code. it now works for function module, programs and classes.

**Comment by marcellourbani** [comment]
@uxkjaer 
I planned to go through this on the weekend but was too lazy :( 
Merged, but plan to do a couple minor changes before releasing it, probably early this week

Marcello

**Comment by uxkjaer** [comment]
Sounds great. Let me know if there is anything I can help with.

File: client/src/adt/sapgui/sapgui.ts [hunk 1]
```diff
@@ -164,34 +214,51 @@ export class SapGui {
...
     if (command.okCode) addParm("DYNP_OKCODE", command.okCode)
-    return `${command.command} ${params}`
+    return `${command.command} ${params} `
   }
 
   private createLauncherContent(command: SapGuiCommand, ticket: string) {
     this.checkConfig()
-    const loginTicket = ticket ? `at="MYSAPSSO2=${ticket}"` : ""
-    const lang = this.language ? `Language=${this.language}` : ""
+    const loginTicket = ticket ? `at = "MYSAPSSO2=${ticket}"` : ""
```
Comment (marcellourbani) [inline client/src/adt/sapgui/sapgui.ts:246]: This broke the sapgui integration: logon ticket not passed correctly

File: client/src/adt/sapgui/sapgui.ts [hunk 2]
```diff
@@ -164,34 +214,51 @@ export class SapGui {
...
-${loginTicket}
-${lang}
-[Function]
-Type=${command.type}
-Command=${this.commandString(command)}
-[Configuration]
-GuiSize=Maximized
-[Options]
-Reuse=1`
+    guiparm = "${this.connectionString}"
```
Comment (marcellourbani) [inline client/src/adt/sapgui/sapgui.ts:249]: ...and this broke the rest.
It's a multiline string, not js code, and the resulting file ends up being syntactically incorrect

File: client/src/commands/registry.ts [hunk 3]
```diff
@@ -4,6 +4,7 @@ export const AbapFsCommands = {
   search: "abapfs.search",
   create: "abapfs.create",
   execute: "abapfs.execute",
+  openInGui: "abapfs.openInGui",
```
Comment (marcellourbani) [inline client/src/commands/registry.ts:7]: renamed this for backwards compatibility

File: package.json [hunk 4]
```diff
@@ -829,6 +833,16 @@
                   "title": "Custom certificate authority certificate",
                   "default": "/secrets/myCA.pem"
                 },
+                "useWebguiEmbedded": {
```
Comment (marcellourbani) [inline package.json:836]: replaced these with a dropdown

**Comment by marcellourbani** [comment]
@uxkjaer I'm working on it, found a few issues and a few things I'd like to change even if not strictly issues (i.e. I restored the previous F5 binding for backwards compatibility).
I will probably push my WIP so you can have a look

**Pull request merged**